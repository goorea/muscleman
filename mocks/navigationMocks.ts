export const navigationDispatchMock = jest.fn();
export const navigationNavigateMock = jest.fn();
export const navigationGoBackMock = jest.fn();
export const navigationSetOptionsMock = jest.fn();
export const navigationSetParamsMock = jest.fn();
export const navigationCanGoBackMock = jest.fn();
export const navigationGetParentMock = jest.fn();
export const navigationGetStateMock = jest.fn();
export const navigationReplaceMock = jest.fn();
export const navigationIsFocusedMock = jest.fn();
export const navigationResetMock = jest.fn();
export const navigationPopMock = jest.fn();
export const navigationPushMock = jest.fn();
export const navigationPopToTopMock = jest.fn();
export const navigationAddListenerMock = jest.fn();
export const navigationRemoveListenerMock = jest.fn();
export const navigationJumpToMock = jest.fn();
export const navigationMock = {
  dispatch: navigationDispatchMock,
  navigate: navigationNavigateMock,
  goBack: navigationGoBackMock,
  setOptions: navigationSetOptionsMock,
  setParams: navigationSetParamsMock,
  canGoBack: navigationCanGoBackMock,
  getParent: navigationGetParentMock,
  getState: navigationGetStateMock,
  isFocused: navigationIsFocusedMock,
  reset: navigationResetMock,
  addListener: navigationAddListenerMock,
  removeListener: navigationRemoveListenerMock,
  replace: navigationReplaceMock,
  pop: navigationPopMock,
  push: navigationPushMock,
  popToTop: navigationPopToTopMock,
  jumpTo: navigationJumpToMock,
};
