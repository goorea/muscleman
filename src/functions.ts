import StackFlashMessage, { Options } from 'react-native-stack-flash-message';

import { WeightSet } from '@src/types';
import { Set, TrainingType } from '@src/types/graphql';

export const flash = (options: Options) => StackFlashMessage.show(options);

export const getTrainingTypeForKorean = (type: TrainingType): string => {
  switch (type) {
    case TrainingType.Lower:
      return '하체';
    case TrainingType.Chest:
      return '가슴';
    case TrainingType.Back:
      return '등';
    case TrainingType.Shoulder:
      return '어깨';
    case TrainingType.Arm:
      return '팔';
    case TrainingType.Abdominal:
      return '복근';
    case TrainingType.Cardiovascular:
      return '유산소';
    default:
      return '기타';
  }
};

export const isWeightSets = (sets: Set[]): sets is WeightSet[] =>
  sets.every(
    set =>
      (set as WeightSet).weight !== undefined &&
      (set as WeightSet).count !== undefined,
  );
