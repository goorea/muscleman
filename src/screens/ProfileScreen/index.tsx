import { BottomTabScreenProps } from '@react-navigation/bottom-tabs';
import { CompositeScreenProps } from '@react-navigation/native';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import dayjs from 'dayjs';
import React, { useCallback } from 'react';
import { useRecoilValue } from 'recoil';

import Button from '@src/components/Button';
import ConfirmModal from '@src/components/ConfirmModal';
import SocialIcon from '@src/components/SocialIcon';
import Text from '@src/components/Text';
import { userState } from '@src/recoils';
import { Gender, User } from '@src/types/graphql';
import { MainTabParamList, RootStackParamList } from '@src/types/navigation';

import useEdit from './hooks/useEdit';
import useEditProfileImage from './hooks/useEditProfileImage';
import useLogout from './hooks/useLogout';
import useWithdrawal from './hooks/useWithdrawal';
import {
  Circle,
  Container,
  Divider,
  EditButton,
  EditProfileImage,
  EmailContainer,
  FooterContainer,
  HeaderContainer,
  Info,
  InfoContainer,
  LoginButton,
  NicknameContainer,
  NicknameEmailContainer,
  NotLoginContainer,
  FooterDivider,
} from './styled';

type P = CompositeScreenProps<
  BottomTabScreenProps<MainTabParamList, 'Profile'>,
  NativeStackScreenProps<RootStackParamList>
>;

const ProfileScreen: React.FC<P> = ({ navigation }) => {
  const user = useRecoilValue<User | undefined>(userState);
  const onLogin = useCallback(
    () =>
      navigation.navigate('Auth', {
        screen: 'Login',
      }),
    [navigation],
  );
  const { editProfileImage, editProfileImageNode } = useEditProfileImage();
  const { edit, editIconProps } = useEdit();
  const { logout } = useLogout();
  const { confirmModalRef, handleConfirm, withdrawal, loading } =
    useWithdrawal();

  if (!user) {
    return (
      <NotLoginContainer>
        <Text weight="bold">Î°úÍ∑∏Ïù∏ ÌõÑ Ïù¥Ïö© Í∞ÄÎä•Ìï©ÎãàÎã§!</Text>
        <Circle>
          <Text size={40}>üí™</Text>
        </Circle>
        <LoginButton
          onPress={onLogin}
          type="outline"
          color="primary"
          title="Î°úÍ∑∏Ïù∏ ÌïòÎü¨Í∞ÄÍ∏∞"
          size={12}
        />
      </NotLoginContainer>
    );
  }

  return (
    <>
      <Container>
        <HeaderContainer>
          <EditProfileImage
            type="clear"
            onPress={editProfileImage}
            node={editProfileImageNode}
          />
          <NicknameEmailContainer>
            <NicknameContainer>
              <Text weight="bold">{user.nickname}</Text>
              <EditButton onPress={edit} type="clear" icon={editIconProps} />
            </NicknameContainer>
            <EmailContainer>
              {user.provider && (
                <>
                  <SocialIcon provider={user.provider} size={18} fill={true} />
                  <Divider />
                </>
              )}
              <Text size={12} color="grey2">
                {user.email}
              </Text>
            </EmailContainer>
          </NicknameEmailContainer>
        </HeaderContainer>

        <InfoContainer>
          <Info>
            <Text weight="bold">{user.name}</Text>
          </Info>
          {!!user.gender && (
            <Info>
              <Text weight="bold">
                {user.gender === Gender.Male ? 'ÎÇ®Ïûê' : 'Ïó¨Ïûê'}
              </Text>
            </Info>
          )}
          {!!user.birth && (
            <Info>
              <Text weight="bold">{dayjs(user.birth).format('YYYYMMDD')}</Text>
            </Info>
          )}
        </InfoContainer>

        {!!user.tel && (
          <InfoContainer>
            <Info>
              <Text weight="bold">{user.tel}</Text>
            </Info>
          </InfoContainer>
        )}

        <FooterContainer>
          <Button
            onPress={logout}
            type="clear"
            title="Î°úÍ∑∏ÏïÑÏõÉ"
            size={12}
            color="grey3"
            weight="normal"
          />
          <FooterDivider size={12} color="grey3">
            |
          </FooterDivider>
          <Button
            onPress={handleConfirm}
            type="clear"
            title="ÌöåÏõêÌÉàÌá¥"
            size={12}
            color="grey3"
            weight="normal"
          />
        </FooterContainer>
      </Container>

      <ConfirmModal
        ref={confirmModalRef}
        message={
          'Ï†ïÎßê ÌöåÏõê ÌÉàÌá¥ÌïòÏãúÍ≤†ÏäµÎãàÍπå?\nÌÉàÌá¥ ÌõÑÏóêÎäî Î≥µÍµ¨Í∞Ä Î∂àÍ∞ÄÎä• Ìï©ÎãàÎã§.'
        }
        onConfirm={withdrawal}
        confirmText="ÌÉàÌá¥"
        loading={loading}
      />
    </>
  );
};

export default ProfileScreen;
