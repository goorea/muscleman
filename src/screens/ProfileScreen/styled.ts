import styled from 'styled-components/native';

import Button from '@src/components/Button';
import Image from '@src/components/Image';
import Text from '@src/components/Text';
import { flexCenter } from '@src/styles/flex';

export const Container = styled.View`
  flex: 1;
  padding: 20px;
`;

export const HeaderContainer = styled.View`
  flex-direction: row;
  align-items: center;
`;

export const EditProfileImage = styled(Button)``;

export const ProfileImage = styled(Image)`
  width: 60px;
  height: 60px;
  border-radius: 30px;
  margin-right: 10px;
`;

export const CameraContainer = styled.View`
  width: 30px;
  height: 30px;
  border-radius: 15px;
  position: absolute;
  bottom: -2px;
  right: 4px;
  background-color: ${({ theme }) => theme.background};
  border: 0.5px solid ${({ theme }) => theme.greyOutline};
  ${flexCenter};
`;

export const NicknameEmailContainer = styled.View`
  flex: 1;
`;

export const NicknameContainer = styled.View`
  flex: 1;
  flex-direction: row;
  align-items: center;
  border-bottom-width: 0.5px;
  border-bottom-color: ${({ theme }) => theme.greyOutline};
`;

export const EditButton = styled(Button)`
  padding: 4px;
`;

export const EmailContainer = styled.View`
  padding: 5px;
  background-color: ${({ theme }) => theme.grey5};
  margin-top: 5px;
  flex-direction: row;
  align-items: center;
`;

export const Divider = styled.View`
  width: 1px;
  height: 12px;
  background-color: ${({ theme }) => theme.greyOutline};
  margin: 0 6px;
`;

export const InfoContainer = styled.View`
  flex-direction: row;
  margin: 10px -6px 0;
`;

export const Info = styled.View`
  flex: 1;
  border-bottom-width: 0.5px;
  border-bottom-color: ${({ theme }) => theme.greyOutline};
  padding: 10px 0;
  margin: 0 6px;
`;

export const FooterContainer = styled.View`
  margin-top: auto;
  flex-direction: row;
  justify-content: flex-end;
`;

export const FooterDivider = styled(Text)`
  margin: 0 3px;
`;
