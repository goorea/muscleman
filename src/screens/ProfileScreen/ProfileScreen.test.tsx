import { navigationMock } from '@mocks/navigationMocks';
import { render } from '@testing-library/react-native';
import React from 'react';

import { wrapper } from '@tests/functions';

import ProfileScreen from './index';

describe('ProfileScreen 컴포넌트', () => {
  const rendered = () =>
    render(
      <ProfileScreen
        navigation={navigationMock}
        route={{
          key: '',
          name: 'Profile',
        }}
      />,
      { wrapper },
    );

  it('렌더링이 올바르게 된다', () => {
    const { toJSON } = rendered();

    expect(toJSON()).toMatchSnapshot();
  });
});
