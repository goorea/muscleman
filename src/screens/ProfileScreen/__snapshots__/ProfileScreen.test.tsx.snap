// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileScreen 컴포넌트 렌더링이 올바르게 된다 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    italic={false}
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "SpoqaHanSansNeo-Bold",
          "fontSize": 16,
          "fontStyle": "normal",
        },
      ]
    }
    weight="bold"
  >
    로그인 후 이용 가능합니다!
  </Text>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2687ED",
          "borderRadius": 50,
          "height": 100,
          "justifyContent": "center",
          "marginTop": 10,
          "width": 100,
        },
      ]
    }
  >
    <Text
      italic={false}
      size={40}
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "SpoqaHanSansNeo-Regular",
            "fontSize": 40,
            "fontStyle": "normal",
          },
        ]
      }
    >
      💪
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#2687ED",
        "borderRadius": 8,
        "borderStyle": "solid",
        "borderWidth": 1,
        "justifyContent": "center",
        "marginTop": 20,
        "opacity": 1,
        "paddingBottom": 14,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 14,
      }
    }
  >
    <Text
      color="primary"
      italic={false}
      size={12}
      style={
        Array [
          Object {
            "color": "#2687ED",
            "fontFamily": "SpoqaHanSansNeo-Bold",
            "fontSize": 12,
            "fontStyle": "normal",
          },
        ]
      }
      weight="bold"
    >
      로그인 하러가기
    </Text>
  </View>
</View>
`;
